src/esp/sim/Simulator.cpp

```diff
@@ -809,5 +821,168 @@ void Simulator::setObjectLightSetup(int objectID,
+  gfx::RenderCamera& renderCamera_ = sceneGraph.getDefaultRenderCamera();
+
+  // try a ray test
+  physics::BulletPhysicsManager* bpm =
+      static_cast<physics::BulletPhysicsManager*>(physicsManager_.get());
+  Magnum::Vector3 cast =
+      (point - renderCamera_.node().absoluteTranslation()).normalized();
+  btCollisionWorld::AllHitsRayResultCallback hit =
+      bpm->castRay(renderCamera_.node().absoluteTranslation(), cast);
+  Magnum::Vector3 hitPoint;
+
+  for (int hitIx = 0; hitIx < hit.m_hitPointWorld.size(); hitIx++) {
+    if (hit.m_hitFractions.at(hitIx) < best_fraction) {
+      best_fraction = hit.m_hitFractions.at(hitIx);
+      
+      nearestObjId = bpm->getObjectIDFromCollisionObject(
+          hit.m_collisionObjects.at(hitIx));
+      hitPoint = Magnum::Vector3{hit.m_hitPointWorld.at(hitIx)};
+    }
+  }
+
+  if ((hitPoint - refPoint).length() > distance) {
+    return -1;
+  }
+
+  return nearestObjId;
+}
+
+void Simulator::grabReleaseObjectUsingCrossHair(Magnum::Vector2i windowSize) {
+  Magnum::Vector2i crossHairPos = Magnum::Vector2i{windowSize * 0.5};
+  Magnum::Vector3 point = unproject(crossHairPos, windowSize, 0);
+  Magnum::Vector3 refPoint = getAgent(0)->node().absoluteTranslation();
+  int nearestObjId = findNearestObjectUnderCrosshair(0, point, refPoint, windowSize);
+
+  auto agentBodyNode_ = &getAgent(0)->node();
+    
+  if (grippedObjectId != -1) {
+    // already gripped, so let it go
+    Magnum::Matrix4 agentT =
+        agentBodyNode_->MagnumObject::transformationMatrix();
+    physicsManager_->setTransformation(
+        grippedObjectId, agentT * gripOffset);
+
+    Magnum::Vector3 position = physicsManager_->getTranslation(grippedObjectId);
+    bool isNav = pathfinder_->isNavigable(Eigen::Map<esp::vec3f>(position.data()));
+
+    //check for collision (apparently this is always true)
+    if (
+      //(physicsManager_->contactTest(grippedObjectId)) && 
+      (!isNav)) {
+      LOG(INFO) << "Colliding with object or position is not navigable";
+      return;
+    }
+    
+    physicsManager_->setObjectMotionType(grippedObjectId,
+        esp::physics::MotionType::STATIC);
+    grippedObjectId = -1;
+  } else if (nearestObjId != -1) {
+    Magnum::Matrix4 agentT =
+        agentBodyNode_->MagnumObject::transformationMatrix();
+    gripOffset = agentT.inverted() *
+        physicsManager_->getObjectSceneNode(nearestObjId)
+        .transformation();
+    physicsManager_->setObjectMotionType(nearestObjId,
+        esp::physics::MotionType::KINEMATIC);
+    grippedObjectId = nearestObjId;
+  } else {
+    return;
+  }
+
+  esp::nav::NavMeshSettings navMeshSettings;
+  navMeshSettings.setDefaults();
+  navMeshSettings.agentRadius = 0.3f;
+
+  recomputeNavMesh(*getPathFinder().get(), navMeshSettings, true);
+  toggleNavMeshVisualization();
+  toggleNavMeshVisualization();
+}
+
+Magnum::Vector3 Simulator::unproject(const Magnum::Vector2i& crosshairPos,
```

Comment:
Maybe we want to support this in the VisualSensor or RenderCamera instead of through Simulator? Currently, there can be multiple sensors in the simulator with different transformations, so there is some ambiguity here.

Reply:
@aclegg3 for moving `unproject` to RenderCamera I am facing issues using it on JS side in `grabReleaseObject` action. When returned object is a reference emscripten uses a copy constructor to instantiate the returned value. `SceneGraph` `SceneNode` and `RenderCamera` don't have a copy constructor and I am not able to expose bindings for those. What would be a better way to implement with the above constraints? Should I expose methods in Simulator to call `unproject`, that way I don't have to bind `RenderCamera` and `SceneGraph` on JS side.